### ISSUE TITLE
Invalid "__iter__" method does not return an object (reportGeneralTypeIssues)

### ISSUE THREAD

**Issue by Kache** [microsoft/pyright#2244]
**Describe the bug**

Incorrectly seeing:

```
error: "Foo" is not iterable
    "__iter__" method does not return an object (reportGeneralTypeIssues)
```

**To Reproduce**
`pyright file.py` of code below

**Expected behavior**
`mypy file.py` doesn't error

**Screenshots or Code**

```python
from __future__ import annotations
from typing import Iterator


class Foo:
    def __iter__(self) -> Iterator[dict] | Iterator[tuple]:
        return self._iterator()

    def _iterator(self):
        yield (1,)


foo = Foo()

for f in foo:  # here
    print(f)
```

**VS Code extension or command-line**
coc.nvim & command line `pyright 1.1.144`

**Comment by erictraut** [comment]
I think the correct return type annotation in this case is `Iterator[dict | tuple]`. If you change it to this, the type error is eliminated.

**Comment by erictraut** [comment]
Upon further thought, I think pyright should accept unions of `Iterator` types as well. I've updated the logic to accommodate this.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.165, which I just published. It will also be included in the next release of pylance.
